Implement scriptID for customEventHooks

HotaruBlaze opened the pull request:
This allows you to easily remove and load active scripts via its scriptID without restarting.

When registering an handler/validator, you can optionally receive an scriptID, tied to your filename.
This can be used to remove the event from tes3mp's event table
This also adds a `"OnScriptUnload"` event that's triggered before we unload that scripts functions

---

- [x] Have working ScriptID/UUID support for scripts being loaded via native methods, Such as require
- [x] Add support for loading/reloading scripts via /load, /unload and /reload
- [x] Fix path differences between `require()` and `/load`, as the ScriptID is based on the path, this will generate two different ID's

---
~~This currently has a dedicated command for unloading due to their being an issue around /load and the filepath not matching what tes3mp is currently using, tho after figuring out this blocking issue, we should be able to implement it correctly into /load~~ 
This is fixed now
```
/load           = .\\/server/scripts/custom/chatUtils.lua
server loading  = .\server\scripts\custom/chatUtils.lua
```

HotaruBlaze committed:
Implement scriptID for customEventHooks

This allows you to easily remove and load active scripts via its scriptID without restarting

HotaruBlaze committed:
Add Validator for OnScriptUnload

HotaruBlaze committed:
Simplify generateScriptID

HotaruBlaze committed:
Remove useless return nil

HotaruBlaze committed:
Dont add os.time() for generateScriptID

magicaldave commented:
Alright, so we've tested this functionality on two debug servers now. It seems to work. Let's please get a list in here of what this MR is missing or cannot do so we can approach completing it.

HotaruBlaze commented:
The only real issue is the paths don't work between tes3mp loading it via require and via /load, because the paths don't match, it'll load the script twice

HotaruBlaze commented:
I added a checklist of the main issues I'm aware of, not gonna address the debug issue until that comes about as that's gonna end up requiring us to add another wrapper or changing how scripts are loaded.

magicaldave committed:
Cleanup(customEventHooks): switch to dreamweave namespace and silence an unused var

magicaldave committed:
Feat(script loader): Add function to normalize path strings for lua conventions
& sandbox user inputs, use it to generate UUIDs for all scripts

magicaldave committed:
Fix(eventHooks): Normalize the path when generating UUIDs instead of whatever I actually did

HotaruBlaze committed:
Add lookup table for scriptID and tweak if statements

HotaruBlaze committed:
Fix loading and add unloading

magicaldave reviewed (approved):
Everything here seems good to go. We'll merge this with the name change.

SaintMercury commented on scripts/commandHandler.lua:
```diff
@@ -715,6 +713,17 @@ function commandHandler.ProcessCommand(pid, cmd)
+                local scriptID = customEventHooks.generateScriptID(scriptName)
+
+                -- Trigger OnScriptUnload event
+                local eventStatus = customEventHooks.triggerValidators("OnScriptUnload", {scriptID})
```
I don't think that the command should be controlling the logic of validation/defaulthandling/handlers.

Isn't there a file where this sort of stuff is already done, and we can just emit an event?

HotaruBlaze replied:
we could try moving it, I just had issues getting it to behave in there, someone else is welcome to take a shot at it.

SaintMercury replied:
What were the problems you were seeing?

HotaruBlaze replied:
I mostly couldn't get the events to actually register and be passed though/triggered

SaintMercury replied:
I'll try giving it a look sometime this week

SaintMercury commented on scripts/commandHandler.lua:
```diff
@@ -731,12 +740,67 @@ function commandHandler.ProcessCommand(pid, cmd)
+                package.loaded[scriptName] = nil
+    
+                Players[pid]:Message(color.GoldenRod  .. scriptName .. color.LightGray .. " has been successfully unloaded.\n".. color.Default)
+            else
```
if we are gonna do early exits, prefer to bring this `else` up, convert to an `if` and remove the previous `if`

SaintMercury commented on scripts/customEventHooks.lua:
```diff
@@ -2,64 +2,194 @@ local customEventHooks = {}
+	end
+	math.randomseed(seed)
+
+    local template = "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"
```
does the template need to be a local variable to the function?

HotaruBlaze replied:
we can probably move it to the start of the file, theirs no reason for this to be a local in the func

SaintMercury commented on scripts/customEventHooks.lua:
```diff
@@ -2,64 +2,194 @@ local customEventHooks = {}
+		return nil
+	end
+
+	if customEventHooks.scriptID.generatedScriptIDs[filePath] then
```
why is generate optionally returning an already loaded script id?

If we are calling generate multiple times, then that's likely a problem.

HotaruBlaze replied:
This is just caching because else we generate a fresh ID every time we add a handler, its designed to have the same ID but multiple validators and handlers registered to one scriptID

SaintMercury commented on scripts/customEventHooks.lua:
```diff
@@ -2,64 +2,194 @@ local customEventHooks = {}
+		return customEventHooks.scriptID.generatedScriptIDs[filePath]
+	end
+
+	local seed = 0
```
seed should not be a function local variable?

HotaruBlaze replied:
seed should be local, its only gonna match up the input and the input can change

SaintMercury replied:
So seed is monotonic per scriptId? it's unlikely but isn't it possible to still have collisions?

Ina  dumb case of : scripts/saint/main and scrtipts/main/saint, won't their normalized path seed be the same?

SaintMercury commented on scripts/customEventHooks.lua:
```diff
@@ -2,64 +2,194 @@ local customEventHooks = {}
 
+
+function customEventHooks.getscriptID(scriptName)
+	local ScriptID = customEventHooks.generateScriptID(scriptName)
```
get shouldn't generate

HotaruBlaze replied:
yeah we can fix this, this was more of a old logic thing that got missed

SaintMercury commented on scripts/customEventHooks.lua:
```diff
@@ -2,64 +2,194 @@ local customEventHooks = {}
+	local scriptIDEvents = customEventHooks.scriptID[scriptID]
+
+	-- If there are no events associated with the scriptID, return
+	if not scriptIDEvents then
```
Verbose log here would be nice if someone goofs up and passes an id for nothing

SaintMercury commented on scripts/customEventHooks.lua:
```diff
@@ -2,64 +2,194 @@ local customEventHooks = {}
+
+	-- Get the events associated with the script ID
+	local scriptIDEvents = customEventHooks.scriptID[scriptID]
+	if not scriptIDEvents then
```
verbose log here would be nice

SaintMercury commented on scripts/customEventHooks.lua:
```diff
@@ -2,64 +2,194 @@ local customEventHooks = {}
+
+function customEventHooks.generateScriptID(filePath)
+	if not string.match(filePath, "%S") then 
+		return nil
```
should it be nil, or throw an error?

What is the likely hood this fails and why would this check fail?

HotaruBlaze replied:
it should never fail, its just more added as a catch incase someone somehow access it directly and feeds  feeds it nothing, its sanitized elsewhere with our code

SaintMercury replied:
I think we SHOULD throw then.

If someone is doing something in a bad way, we should try to fail early before we dive down a rabbit hole we shouldn't let them dig

SaintMercury commented on scripts/customEventHooks.lua:
```diff
@@ -2,64 +2,194 @@ local customEventHooks = {}
+  end
+
+  if not customEventHooks.scriptID[scriptID] then
+    dreamweave.LogMessage(enumerations.log.INFO, string.format('[customEventHooks]: Registered ScriptID "%s" for script "%s"', scriptID, filePath))
```
won't this only ever print once per script but for every event?

Should this not be outside the if like the other logMessage?

HotaruBlaze replied:
This only prints for the first registered instance of that ScriptID, any others are only visible with verbose logging

SaintMercury commented on scripts/commandHandler.lua:
```diff
@@ -731,12 +740,67 @@ function commandHandler.ProcessCommand(pid, cmd)
+            Players[pid]:Message(color.LightGray .. "Use /unloadbyid <scriptID>\n" .. color.Default)
+        else
+            -- Trigger OnScriptUnload event
+            local eventStatus = customEventHooks.triggerValidators("OnScriptUnload", {scriptID})
```
this code should live in `eventHandler.lua`

SaintMercury commented on scripts/commandHandler.lua:
```diff
@@ -731,12 +740,67 @@ function commandHandler.ProcessCommand(pid, cmd)
+            if wasLoaded then
+                local scriptID = customEventHooks.generateScriptID(scriptName)
+    
+                -- Trigger OnScriptUnload event
```
Most of this code should live in `eventHandler.lua`

HotaruBlaze committed:
script(review): customEventHooks changes

HotaruBlaze commented:
~~Don't merge this yet, theirs a bug I'm currently diagnosing~~

HotaruBlaze committed:
Fix commands always generating a new ScriptID

HotaruBlaze commented:
Fixed it, it was a bug from when we removed the lookup table from `customEventHooks.generateScriptID()` and made it always generate a new scriptID with the fixed seed logic.

HotaruBlaze committed:
Warn and detect nested includes triggering event registration

HotaruBlaze committed:
Add nil checks for validators and handlers

HotaruBlaze commented:
We need to add an OnScriptLoad validator and handler, as well as move this logic into eventHandler.

This is now blocked until its added, as many scripts depend on OnServerInit or OnServerPostInit and theirs currently no way for the script itself to know it was loaded.

magicaldave committed:
Fix(scriptID): Add, and trigger, a new "OnScriptLoad" event when scripts are reloaded.

magicaldave committed:
Refactor(ScriptID): Generalize unloadHandlers and unloadValidators into the same function

magicaldave committed:
Merge branch 'master' into i-can-haz-ids
